### PR TITLE
full support for prepend vectorization

### PR DESCRIPTION
**Pull request by campsd**
This adds complete support for prepending vectorization to:

- FF loss functions
- vectorization.view 
- factorize
- backends

The following code now works correctly:

```python
A = ff.tensor('A', 3, 2)
B = ff.tensor('B', 2, 3)
i, j, k = ff.indices('i, j, k')
#tsrex = A[[*i, *j]] * B[[i, j]]
tsrex = A[[i, j]] * B[[j, k]]
target = ff.Factorization.from_tsrex(tsrex)
target = target()
fac = ff.factorize(tsrex, target, nvec=4, append=False)
fac()
```

### CODE REVIEW

File: funfact/algorithm.py [hunk 1]
```diff
@@ -92,10 +92,10 @@ class _Factorization(Factorization, ab.AutoGradMixin):
                 'funfact.loss.'
             )
     try:
-        loss(target, target, **kwargs)
+        loss(target, target, append=append, **kwargs)
     except Exception as e:
         raise AssertionError(
-            f'The given loss function does not accept two arguments:\n{e}'
+            f'The given loss function does not accept three arguments:\n{e}'
```
Comment (yhtang) [inline funfact/algorithm.py:98]: The 'two' arguments here refer to the target and approximate tensors that the loss function is meant to compare. Any additional arguments are 'meta/hyper-parameters' and hence does not count towards the number of arguments that a loss function accepts. 😁

File: funfact/loss.py [hunk 2]
```diff
@@ -12,14 +12,15 @@ def _loss(self, model, target):
...
     def __call__(self, model, target, reduction='mean', sum_vec=True,
-                 **kwargs):
-        if target.ndim == model.ndim - 1:
-            if model.shape[:-1] != target.shape:
-                raise ValueError(f'Target shape {target.shape} and '
-                                 f'model shape {model.shape[:-1]} mismatch.')
-            data_axis = tuple(i for i in range(target.ndim))
-            target = target[..., None]
-        elif target.ndim == model.ndim:
+                 append=True, **kwargs):
```
Comment (yhtang) [inline funfact/loss.py:15]: I find the meaning of `append` a bit cryptic here. This is different from the situation of the `vectorize` method, whose name already provides the context. Here, the context is 'loss', and hence it is not very intuitive as to what 'append' would mean. What about `vectorizes_along_last` or something along the aline?

File: funfact/vectorization.py [hunk 3]
```diff
@@ -51,17 +51,21 @@ def view(fac, tsrex_scalar, instance: int):
             The original, un-vectorized tensor expression.
         instance (int):
             Index along the vectorization dimension.
+        append (bool):
```
Comment (yhtang) [inline funfact/vectorization.py:54]: While I understand it is better to reuse the `append` keyword to ensure API consistency here, there is a similar semantic ambiguity problem as above in the loss function.

At least, we can change this docstring line to say something like 'indicates whether the vectorization dimension was appended or prepended'.

**Review by yhtang — CHANGES_REQUESTED** [review]

**Comment by yhtang** [comment]
Another issue here is what default values shall we assign to `append`. I recall that we want to be default-compatible with NumPy. In that case, perhaps we should make all defaults to `append=False`?

File: funfact/algorithm.py (hunk 1 above)
Comment (campsd) [reply funfact/algorithm.py:98]: Changed back.

File: funfact/vectorization.py (hunk 3 above)
Comment (campsd) [reply funfact/vectorization.py:54]: To have a consistent API, we can also call the argument 'vectorize_along_last' everywhere? Slight difference in meaning when it is an active action ('vectorize`) or details about input (`loss`) but maybe that's fine?

File: funfact/loss.py (hunk 2 above)
Comment (campsd) [reply funfact/loss.py]: Changed here to `vectorized_along_last`.

File: funfact/vectorization.py (hunk 3 above)
Comment (campsd) [reply funfact/vectorization.py:54]: I have updated the doc string, but kept the name.

**Comment by campsd** [comment]
> Another issue here is what default values shall we assign to `append`. I recall that we want to be default-compatible with NumPy. In that case, perhaps we should make all defaults to `append=False`?

Good point, all the defaults should now be False everywhere.

**Review by yhtang — APPROVED** [review]
